### PR TITLE
1180 Guide for use of Speculative Execution

### LINKED ISSUE #1180: Create a Guide to using the `speculative_exec` endpoint
The `speculative_exec` endpoint allows developers to better estimate costs for their Deploys. While it is covered in the JSON-RPC API docs, it would be good to have a more plain-speak tutorial that outlines uses and best practices.

Should be placed in the "Building dApps" section of the docs.

### PR DESCRIPTION
**Pull request by ACStoneCL**
### What does this PR fix/introduce?
This PR introduces a guide for use of the `speculative_exec` endpoint in the Casper client.

Closes #1180 

### Additional context
[Create a Guide to using the speculative_exec endpoint #1180](https://github.com/casper-network/docs/issues/1180)

### Checklist

- [x] Docs are successfully building - `yarn install && yarn run build`.
- [x] For new **internal** links I used *relative file paths* (with .md extension) - e.g. `../../faq/faq-general.md` - instead of introducing *absolute file path*, or *relative/absolute URL*.
- [x] My changes follow the [Casper docs style guidelines](https://docs.casper.network/resources/contribute-to-docs/).
- [x] All technical procedures have been tested (if you want help with this, mention it in [Reviewers](#reviewers)).

### Reviewers
Editing: @ipopescu / Tech @bpr or assignee.

### CODE REVIEW

**Comment by ACStoneCL** [comment]
@bpr I would previously have tagged Karan for tech review, but if you have someone you would prefer review this, I can tag them instead.

**Review by ipopescu — COMMENTED** [review]
Overall it looks good, but I have a couple of questions before approving.

File: source/docs/casper/developers/dapps/speculative-exec.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      }
+    }
+  }
+}
+
+```
+
+:::note
+
+Cost estimates acquired through `speculative_exec` may vary from the cost of sending the same Deploy to a Casper network. Speculative execution is a tool to help narrow down the potential cost of sending a Deploy, but many factors can cause the actual cost to vary.
```
[the remark this answers is not in the document]
Comment (ACStoneCL) [reply source/docs/casper/developers/dapps/speculative-exec.md:49]: The impression I get is that `speculative_exec` can only ever give you a rough estimate, so we need to be clear that even on the same network, the actual Deploy may vary in cost from the speculative Deploy.

[the remark this answers is not in the document]
Comment (ipopescu) [reply source/docs/casper/developers/dapps/speculative-exec.md:49]: Ok. I am still curious to know what would cause these variations though.

**Review by ipopescu — APPROVED** [review]
LGTM! I'm still curious about what would cause variation in cost estimates. Perhaps @bpr can give us some insight.

File: source/docs/casper/developers/dapps/speculative-exec.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (bpr) [reply source/docs/casper/developers/dapps/speculative-exec.md:49]: One reason, that I'm sure you're familiar with, is that load on the network may vary, which will affect how the network processes any given request. Network performance always varies, even when the network doesn't have a varying external load.  If memory is being used, there will be time that goes into cleaning up or compacting, which will occur at unpredictable times. In systems analysis we have to use probabilistic and statistical methods to deal with this non-determinism, so we'll always have performance metrics with some variance.

Does that help?

**Review by bpr — APPROVED** [review]
LGTM

File: source/docs/casper/developers/dapps/speculative-exec.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ipopescu) [reply source/docs/casper/developers/dapps/speculative-exec.md:49]: Yes, thanks @bpr!

**Pull request merged**